velox/exec/tests/RowNumberTest.cpp

```diff
@@ -218,4 +275,40 @@ TEST_F(RowNumberTest, largeInput) {
   testLimit(5'000);
 }
 
+TEST_F(RowNumberTest, reclaimFail) {
+  auto spillDirectory = exec::test::TempDirectoryPath::create();
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>({1, 2, 1, 2, 1, 2, 1}),
+      makeFlatVector<int64_t>({1, 2, 3, 4, 5, 6, 7}),
+  });
+
+  createDuckDbTable({data});
+  setupMemory(1 * MB);
+  try {
+    AssertQueryBuilder(duckDbQueryRunner_)
+        .config(core::QueryConfig::kSpillEnabled, "true")
+        .config(core::QueryConfig::kRowNumberSpillEnabled, "true")
+        .spillDirectory(spillDirectory->path)
+        .queryCtx(newQueryCtx(16 * MB))
+        .plan(PlanBuilder(std::make_shared<core::PlanNodeIdGenerator>())
+                  .values({data, data, data})
+                  .rowNumber({"c0"}, 2, false)
+                  .project({"c0", "c1"})
+                  .planNode())
+        .assertResults(
+            "SELECT c0, c1 "
+            "FROM ("
+            " SELECT *, row_number() over (partition by c0) AS rn FROM tmp) "
+            "WHERE rn <= 2");
+  } catch (const VeloxException& e) {
```

Comment:
Use VELOX_ASSERT_THROW and if needed add a similar macro for memory-specific exceptions. This pattern is buggy because the test allows the query to succeed. You would need to add VELOX_FAIL in the 'try' body to fix that, but people always forget.

Reply:
@mbasmanova Yes, this version is buggy and I found it is not guaranteed to trigger the 'hasSpilled' check logic added in this PR.

I've updated this test (**`RowNumberTest.reclaimFail`**), it is designed to make `RowNumber::maybeReserve` spill through `RowNumber::reclaim` and return false to guard the 'hasSpilled' check logic added in this PR. It creates multiple tasks to trigger the memory reclaim, some tasks may fail with memory-related exceptions, which is acceptable and some may not, so I catch the expected exception with the expected error code and rethrow the unexpected exception. But it is kind of complicated and over design to just test the newly added logic in this PR, and it introduces duplicate methods `setupMemory` and `newQueryCtx`.

So I also added another way to test  (**`RowNumberTest.reclaimFail`**), the 'hasSpilled' check logic added in this PR. It injected the test method into the production code and triggered the spill no matter whether ``RowNumber::maybeReserve` triggered or not. It does not need the `setupMemory`, `newQueryCtx` method.

What do you think, and which one do you prefer? 

cc @xiaoxmeng